rtsf-at-checkout-device-scale/driver/scaleDriver_test.go

```diff
@@ -0,0 +1,279 @@
+				return
+			}
+			require.NoError(t, err)
+			require.NotNil(t, gotRes)
+		})
+	}
+}
+
+func Test_findSerialPort(t *testing.T) {
+
+	tests := []struct {
+		name     string
+		portInfo enumerator.PortDetails
+		want     string
+		wantErr  bool
+	}{
+		{
+			name: "valid case",
+			portInfo: enumerator.PortDetails{
+				Name:         "testDevice",
+				IsUSB:        true,
+				PID:          "6001",
+				VID:          "0403",
+				SerialNumber: "0123456",
+			},
+			want:    "testDevice",
+			wantErr: false,
+		},
+		{
+			name: "vid pid not found",
+			portInfo: enumerator.PortDetails{
+				Name:         "testDevice",
+				IsUSB:        true,
+				PID:          "0000",
+				VID:          "0000",
+				SerialNumber: "0123456",
+			},
+			want:    "",
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var ports []*enumerator.PortDetails
+			ports = append(ports, &tt.portInfo)
+			got, err := findSerialPort(ports, "6001", "0403")
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+			assert.Equal(t, tt.want, got)
+		})
+	}
+}
+
+func TestScaleDriver_AddDevice(t *testing.T) {
+	config := scale.Config{
+		PortName:        "/dev/tty.usbserial-test",
+		BaudRate:        9600,
+		DataBits:        7,
+		StopBits:        1,
+		MinimumReadSize: 1,
+		ParityMode:      2,
+		TimeOutMilli:    500,
+	}
+
+	testDevice := scale.InitializeMockDevice(&config)
+
+	type args struct {
+		deviceName string
+		protocols  map[string]models.ProtocolProperties
+		adminState models.AdminState
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "no serial protocol",
+			args: args{
+				deviceName: "testDeviceName",
+				protocols:  nil,
+				adminState: "full",
+			},
+			wantErr: true,
+		},
+		{
+			name: "no pid",
+			args: args{
+				deviceName: "testDeviceName",
+				protocols: map[string]models.ProtocolProperties{
+					"serial": {
+						"VID": "0403",
+					},
+				},
+				adminState: "full",
+			},
+			wantErr: true,
+		},
+		{
+			name: "no vid",
+			args: args{
+				deviceName: "testDeviceName",
+				protocols: map[string]models.ProtocolProperties{
+					"serial": {
+						"PID": "6001",
+					},
+				},
+				adminState: "full",
+			},
+			wantErr: true,
+		},
+		{
+			name: "port list can not be found",
+			args: args{
+				deviceName: "testDeviceName",
+				protocols: map[string]models.ProtocolProperties{
+					"serial": {
+						"VID": "0403",
+						"PID": "6001",
+					},
+				},
+				adminState: "full",
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			drv := getDefaultScaleDriver()
+			drv.scaleDevice = &scaleDevice{
+				serialDevice: testDevice,
+			}
+
+			if err := drv.AddDevice(tt.args.deviceName, tt.args.protocols, tt.args.adminState); (err != nil) != tt.wantErr {
+				t.Errorf("ScaleDriver.AddDevice() error = %v, wantErr %v", err, tt.wantErr)
+			}
```

Comment:
Should never need to use `t.Errorf`. If used then refactor.

```suggestion
			err := drv.AddDevice(tt.args.deviceName, tt.args.protocols, tt.args.adminState);
			if  tt.wantErr {
				require.Error(..)
			}
			
			require.NoError(..)
```